typescript/ccip-server/src/services/CallCommitmentsService.ts

```diff
@@ -0,0 +1,293 @@
+    commitment: string,
+    revealMessageId: string,
+  ) {
+    console.log('Fetching commitment from DB', commitment, revealMessageId);
+    const record = await prisma.commitment.findUnique({
+      where: { commitment_revealMessageId: { commitment, revealMessageId } },
+    });
+    if (!record) {
+      console.log('Commitment not found in DB', commitment, revealMessageId);
+      throw new Error('Commitment not found');
+    }
+    return record;
+  }
+
+  // Validate the commitment events by checking the transaction receipt
+  // and parsing the events emitted by the InterchainAccountRouter.
+  // This ensures that the commitment was dispatched correctly and
+  // return the ICA address.
+  // Throws if validation fails.
+  private async validateCommitmentEvents(
+    data: z.infer<typeof postCallsSchema>,
+    commitment: string,
+  ): Promise<{ ica: string; revealMessageId: string }> {
+    const provider = this.multiProvider.getProvider(data.originDomain);
+    const receipt = await provider.getTransactionReceipt(
+      data.commitmentDispatchTx,
+    );
+    if (!receipt)
+      throw new Error(
+        `Transaction not found: ${data.commitmentDispatchTx} on domain ${data.originDomain}`,
+      );
+
+    // 2) Verify reveal event
+    this.ensureCommitmentDispatched(receipt, commitment);
+
+    // 3) Extract reveal message ID
+    const revealMessageId = this.extractRevealMessageId(receipt);
+
+    // 4) Derive ICA from RemoteCallDispatched
+    const ica = await this.deriveIcaFromRemoteCallDispatched(
+      receipt,
+      data.originDomain,
+    );
+    return { ica, revealMessageId };
+  }
+
+  /**
+   * Ensure a CommitRevealDispatched event matching the commitment exists.
+   */
+  private ensureCommitmentDispatched(receipt: any, commitment: string): void {
+    const iface = InterchainAccountRouter__factory.createInterface();
+    const revealTopic = iface.getEventTopic('CommitRevealDispatched');
+    const revealLogs = receipt.logs.filter(
+      (l: any) => l.topics[0] === revealTopic,
+    );
+    if (revealLogs.length === 0) {
+      throw new Error('CommitRevealDispatched event not found');
+    }
+    const matched = revealLogs
+      .map((l: any) => iface.parseLog(l))
+      .some((parsed: any) => parsed.args.commitment === commitment);
+    if (!matched) {
+      throw new Error('No matching CommitRevealDispatched for this commitment');
+    }
+  }
+
+  /**
+   * Parse the RemoteCallDispatched event from the receipt and derive the ICA address.
+   */
+  private async deriveIcaFromRemoteCallDispatched(
+    receipt: TransactionReceipt,
+    originDomain: number,
+  ): Promise<string> {
+    const iface = InterchainAccountRouter__factory.createInterface();
+    const callTopic = iface.getEventTopic('RemoteCallDispatched');
+    const callLog = receipt.logs.find((l) => l.topics[0] === callTopic);
+    if (!callLog) throw new Error('RemoteCallDispatched event not found');
+    const parsedCall = iface.parseLog(callLog);
```

Comment:
this seems like it should also be done "in context"

Reply:
parsing all of the logs in a single scan of the tx receipt

Reply:
Consider moving `ensureCommitmentDispatched` into `extractRevealMessageId`

Reply:
c5c3d6416